### DISCUSSION TITLE
User defined literals operator.

### DISCUSSION THREAD

**Discussion by nathan130200** [dotnet/csharplang#6705] [Language Ideas]
A nice feature that would benefit C# is literal operators like C++. The programmer can define specific literal operators, like u8 present in .NET 7 / C# 11

`this_thread::sleep_for(1ms)` -> `1ms` -> `std::literals::chrono_literals::operator "" ms(unsigned long long value)`

###
- From: 
    `await Task.Delay(TimeSpan.FromSeconds(1d));`
- To:
    `await Task.Delay(1sec); // user defined literal: sec `

### Syntax:
- `[modifier] static literal operator "[prefix]" "[suffix]" <return type>(argument)`

    Or this one is better syntax:
        `[modifier] static <return type> literal operator "prefix" "suffix" (argument)`

    - Where:
    - `[modifier]` common method modifiers: internal, public, protectd, protected internal
    - `[prefix]` OPTIONAL: Literal operator prefix.
    - `[suffix]` OPTIONAL: Literal operator suffix.
    - Requirements: One of both must be present: prefix or suffix.

### Syntax Sample:
```cs
struct TimeSpan
{
 // ... 

  public static literal operator "" "ms" TimeSpan(double value)
    => TimeSpan.FromMilliseconds(value);
}
```

```cs
[StructLayout(LayoutKind.Explicit)]
struct WideChar {
  [FieldOffset(0)]
  char value;

  [FieldOffset(0)]
  byte b1;

  [FieldOffset(1)]
  byte b2;

  public static literal operator "L" "" WideChar(char c)
    => new() { value = c }; 

  public static implicit operator WideChar(char c)
    => new() { value = c }; 
};

class WideString {
  private WideChar[] buf;
 
   // etc...

  public static literal operator "L" "" WideString(string str)
    => new() { buf = str.Select(x => (WideChar)x) }; 
};


var wideStr = L"two byte, wide string"; // implicit literal: L(string)
var wideCh = L'c'; // implicit literal: L(char)

```

We can call: 

- Auto detect: to TimeSpan operator:
    `var timeToWait = 250ms;`

- Explicit literal operator, when exists more than one with same operator imported in many namespaces
    `TimeSpan timeToWait = 250ms;`

### Notes:

- `ByteSize buffSize = 1024b;` -> User defined literal operator from `ByteSize` struct, will not match with `b` from binary suffix, since we explicit defined.

- OPTIONAL: Common types that uses numeric suffixes like `b` for binary, `d` for decimal, `0x` prefix for hexadecima, `f` for floats, `m` for decimal. Can be integrated into literals operator now with Generic Number in .NET 7

- OPTIONAL: `u8` suffix now in C# 11 can be integrated with literals operator, as builtin operator.

- Literals can have overloads and compiler will decided who is best overload. Except some situations when explicit type is required (for example, two overloads: float and double, specify by `f` or `d` prefix)

**Comment by pelletburner** [comment]
I like the idea, however, this can easily be done with code like this:

`
public static TimeSpan Seconds(this double seconds)
{
  return TimeSpan.FromSeconds(seconds);
}
`

Example usage:

`
  await Task.Delay((1.0d).Seconds);
`

Or simplified by allowing for uint, ulong, etc...

**Reply by nathan130200** [reply]
Oh yes, I just used it as an example with TimeSpan because it's simpler to explain the idea.
